processor/lsmintervalprocessor/internal/merger/value.go

```diff
@@ -0,0 +1,796 @@
+			for k := 0; k < metrics.Len(); k++ {
+				metric := metrics.At(k)
+				metricID := identity.OfMetric(smID, metric)
+				s.metricLookup[metricID] = metric
+
+				switch metric.Type() {
+				case pmetric.MetricTypeSum:
+					dps := metric.Sum().DataPoints()
+					for l := 0; l < dps.Len(); l++ {
+						dp := dps.At(l)
+						streamID := identity.OfStream(metricID, dp)
+						s.numberLookup[streamID] = dp
+					}
+				case pmetric.MetricTypeSummary:
+					dps := metric.Summary().DataPoints()
+					for l := 0; l < dps.Len(); l++ {
+						dp := dps.At(l)
+						streamID := identity.OfStream(metricID, dp)
+						s.summaryLookup[streamID] = dp
+					}
+				case pmetric.MetricTypeHistogram:
+					dps := metric.Histogram().DataPoints()
+					for l := 0; l < dps.Len(); l++ {
+						dp := dps.At(l)
+						streamID := identity.OfStream(metricID, dp)
+						s.histoLookup[streamID] = dp
+					}
+				case pmetric.MetricTypeExponentialHistogram:
+					dps := metric.ExponentialHistogram().DataPoints()
+					for l := 0; l < dps.Len(); l++ {
+						dp := dps.At(l)
+						streamID := identity.OfStream(metricID, dp)
+						s.expHistoLookup[streamID] = dp
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// MergeMetric adds a metric with a provided resource metric and scope
+// metric. Note that overflows during addition will be applied as per
+// the specifications for overflow handling.
+func (v *Value) MergeMetric(
+	rm pmetric.ResourceMetrics,
+	sm pmetric.ScopeMetrics,
+	m pmetric.Metric,
+) error {
+	resID, err := v.addResourceMetrics(rm)
+	if err != nil {
+		return err
+	}
+	scopeID, err := v.addScopeMetrics(resID, sm)
+	if err != nil {
+		return err
+	}
+	v.mergeMetric(resID, scopeID, m)
+	return nil
+}
+
+// Finalize finalizes all overflows in the metrics to prepare it for
+// harvest. This method must be called only once for harvest.
+func (s *Value) Finalize() (pmetric.Metrics, error) {
+	// At this point we need to assume that the metrics are returned
+	// as a final step in the store, thus, prepare the final metric.
+	// In the final metric we have to add datapoint limits.
+	for _, sm := range s.scopeLookup {
+		if !sm.datapointsLimits.HasOverflow() {
+			continue
+		}
+		// Add overflow metric to the scope
+		overflowMetric := sm.ScopeMetrics.Metrics().AppendEmpty()
+		overflowMetric.SetName(overflowMetricName)
+		overflowMetric.SetDescription(overflowMetricDesc)
+		overflowSum := overflowMetric.SetEmptySum()
+		overflowSum.SetAggregationTemporality(pmetric.AggregationTemporalityDelta)
+		overflowDP := overflowSum.DataPoints().AppendEmpty()
+		if err := decorate(
+			overflowDP.Attributes(),
+			s.scopeDPLimitCfg.Overflow.Attributes,
+		); err != nil {
+			return pmetric.Metrics{}, fmt.Errorf("failed to finalize merged metric: %w", err)
+		}
+		overflowDP.SetIntValue(int64(sm.datapointsLimits.EstimateOverflow()))
+	}
+	// Remove any hanging resource or scope which failed to have any entries
+	// due to children reaching their limits.
```

Comment:
Seeing as the scope limit is per resource, and the metric/dp limit is per scope, when would we add an empty resource or scope?

Reply:
This happens due to metric overflows, I have added a comment to explain this better:

```
	// Remove any hanging metrics, scope, or resource which failed to have any
	// entries due to datapoints overflowing. Overflowing datapoints discards
	// that metric and creates a new overflow metric which might result in the
	// original metric and its parent to exist without any datapoints.
```